src/Data/Foreign/Generic/Enum.purs

```diff
@@ -0,0 +1,77 @@
+module Data.Foreign.Generic.EnumEncoding where
+
+import Prelude
+
+import Control.Alt ((<|>))
+import Data.Foreign (F, Foreign, ForeignError(..), fail, readString, toForeign)
+import Data.Foreign.Generic.Types (Options)
+import Data.Generic.Rep (class Generic, Constructor(Constructor), NoArguments(NoArguments), Product, Rec, Sum(Inr, Inl), from, to)
+import Data.Symbol (class IsSymbol, SProxy(..), reflectSymbol)
+import Partial.Unsafe (unsafeCrashWith)
+
+genericDecodeEnum
+  :: forall a rep
+   . Generic a rep
+  => GenericDecodeEnum rep
+  => Options
+  -> Foreign
+  -> F a
+genericDecodeEnum opts = map to <<< decodeEnum opts
+
+genericEncodeEnum
+  :: forall a rep
+   . Generic a rep
+  => GenericEncodeEnum rep
+  => Options
+  -> a
+  -> Foreign
+genericEncodeEnum opts = encodeEnum opts <<< from
+
+class GenericDecodeEnum a where
+  decodeEnum :: Options -> Foreign -> F a
+
+class GenericEncodeEnum a where
+  encodeEnum :: Options -> a -> Foreign
+
+instance sumGenericDecodeEnum
+  :: (GenericDecodeEnum a, GenericDecodeEnum b)
+  => GenericDecodeEnum (Sum a b) where
+  decodeEnum opts f = Inl <$> decodeEnum opts f <|> Inr <$> decodeEnum opts f
+
+instance ctorNoArgsGenericDecodeEnum
+  :: (IsSymbol name)
+  => GenericDecodeEnum (Constructor name NoArguments) where
+  decodeEnum {constructorTagTransform} f = do
+    tag <- readString f
+    unless (constructorTagTransform tag == ctorName) $
+      fail (ForeignError ("Expected " <> show ctorName <> " tag for unary constructor literal " <> ctorName))
+    pure $ Constructor NoArguments
+    where
+      ctorName = constructorTagTransform $ reflectSymbol (SProxy :: SProxy name)
+
+instance sumGenericEncodeEnum
+  :: (GenericEncodeEnum a, GenericEncodeEnum b)
+  => GenericEncodeEnum (Sum a b) where
+  encodeEnum opts (Inl a) = encodeEnum opts a
+  encodeEnum opts (Inr b) = encodeEnum opts b
+
+instance ctorNoArgsGenericEncodeEnum
+  :: (IsSymbol name)
+  => GenericEncodeEnum (Constructor name NoArguments) where
+  encodeEnum {constructorTagTransform} _ = toForeign ctorName
+    where
+      ctorName = constructorTagTransform $ reflectSymbol (SProxy :: SProxy name)
+
+instance productGenericEncodeEnum
+  :: Fail """
+    genericEncodeEnum cannot be used with Product types, as it is meant to work with string literal "Enums", Sum types of no-argument constructors.
+  """
+  => GenericEncodeEnum (Product a b) where
+  encodeEnum _ _ = unsafeCrashWith "unreachable encodeEnum was reached."
+
+instance recGenericEncodeEnum
+  :: Fail """
```

Comment:
Why not just say "doesn't work with constructors with arguments" though? Records and products are both types of arguments. Do you not think it's clear enough?